src/core/libraries/videodec/videodec2_impl.cpp

```diff
@@ -0,0 +1,140 @@
+// SPDX-FileCopyrightText: Copyright 2024 shadPS4 Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "videodec2_impl.h"
+
+#include "common/alignment.h"
+#include "common/assert.h"
+#include "common/logging/log.h"
+
+namespace Libraries::Vdec2 {
+
+static inline void CopyNV12Data(u8* dst, const AVFrame& src) {
+    const auto width = Common::AlignUp(u32(src.width), 16);
+    const auto height = Common::AlignUp(u32(src.height), 16);
+
+    if (src.width == width) {
+        std::memcpy(dst, src.data[0], src.width * src.height);
+        std::memcpy(dst + src.width * height, src.data[1], (src.width * src.height) / 2);
+    } else {
+        const auto luma_dst = dst;
+        for (u32 y = 0; y < src.height; ++y) {
+            std::memcpy(luma_dst + y * width, src.data[0] + y * src.width, src.width);
+        }
+        const auto chroma_dst = dst + width * height;
+        for (u32 y = 0; y < src.height / 2; ++y) {
+            std::memcpy(chroma_dst + y * (width / 2), src.data[0] + y * (src.width / 2),
+                        src.width / 2);
+        }
+    }
+}
+
+Videodec2::Videodec2(const SceVideodec2DecoderConfigInfo& configInfo,
+                     const SceVideodec2DecoderMemoryInfo& memoryInfo) {
+    ASSERT(configInfo.codecType == 1); /* AVC */
+
+    const AVCodec* codec = avcodec_find_decoder(AV_CODEC_ID_H264);
+
+    mCodecContext = avcodec_alloc_context3(codec);
+    mCodecContext->width = configInfo.maxFrameWidth;
+    mCodecContext->height = configInfo.maxFrameHeight;
+
+    avcodec_open2(mCodecContext, codec, nullptr);
+}
+
+Videodec2::~Videodec2() {
+    avcodec_free_context(&mCodecContext);
+    sws_freeContext(mSwsContext);
+}
+
+s32 Videodec2::Decode(const SceVideodec2InputData& inputData, SceVideodec2FrameBuffer& frameBuffer,
+                      SceVideodec2OutputInfo& outputInfo) {
+    AVPacket* packet = av_packet_alloc();
+
+    frameBuffer.isAccepted = false;
+    outputInfo.isValid = false;
+    outputInfo.isErrorFrame = true;
+    outputInfo.pictureCount = 0;
+
+    packet->data = (u8*)inputData.pAuData;
+    packet->size = inputData.auSize;

```

Comment:
You can set `packet->pts` to `inputData.ptsData` and `packet->dts` to `inputData.dtsData`